statsd/statsd.go

```diff
@@ -494,10 +502,16 @@ func (c *Client) Close() error {
 	if c == nil {
 		return ErrNoClient
 	}
+	c.Lock()
 	select {
-	case c.stop <- struct{}{}:
+	case <-c.stop:
+		c.Unlock()
+		return nil
 	default:
 	}
+	c.Unlock()
```

Comment:
I am worried about this locking mechanism. It is still possible that one goroutine takes hold of the lock while another one takes the default branch? Should we maybe either keep it simple (no locks and risk a highly unlikely race where someone calls Close in parallel multiple times) or lock the entire call?

Reply:
We can't lock the entire `Close` call, because `c.Flush()` needs to be able to obtain the lock. It *is* possible one goroutine takes control of the lock while another is past it, but that's OK. ~~The way the lock is used ensures two goroutines can't both close the channel.~~  Actually I need to move `c.Unlock()` after `close(c.stop)`. Then it will ensure two can't both close the channel. (oops)

This is why I had the `shutdown` function, so I could lock the part of this logic that needed to be locked and `defer c.Unlock()` to avoid the risky and more complicated manual unlocking. 

I don't think it's wise to introduce a race that causes a panic, no matter how unlikely, if we can avoid it.